pulp_container/app/authorization.py

```diff
@@ -92,32 +118,70 @@ def _split_into_encoded_groups(self, encoded_base32):
                 kid += char
         return kid
 
-    def determine_access(self, user, scope):
+    def determine_access(self):
         """
         Determine access permissions for a corresponding user.
 
         This method determines whether the user has a valid access permission or not.
-        The determination is based on role based access control. For now, the access
-        is given out to anybody because the role based access control is not implemented
-        yet.
-
-        Args:
-            user (str): A name of the user who is trying to access a registry.
-            scope (str): A requested scope.
+        The determination is based on role based access control.
 
         Returns:
             list: An intersected set of the requested and the allowed access.
 
         """
-        typ, name, actions = scope.split(":")
-        actions_list = actions.split(",")
-        permissions = {"pull"}
-        if user == "admin":
-            permissions.add("push")
-        permitted_actions = list(set(actions_list).intersection(permissions))
-        return {"type": typ, "name": name, "actions": permitted_actions}
-
-    def _generate_claim_set(self, issuer, issued_at, subject, audience, access):
+        typ, name, actions = self.scope.split(":")
+        actions = set(actions.split(","))
+
+        permitted_actions = set()
+        for action in actions:
+            has_permission = self.actions_permissions[action](name)
+            if has_permission:
+                permitted_actions.add(action)
+
+        return {"type": typ, "name": name, "actions": list(permitted_actions)}
+
+    def has_pull_permissions(self, path):
+        """
+        Check if the user has permissions to pull from the repository specified by the path.
+        """
+        if self.user.has_perm("container.view_containerdistribution"):
+            return True
+
+        try:
+            distribution = ContainerDistribution.objects.get(base_path=path)
+        except ContainerDistribution.DoesNotExist:
+            return False
+        if self.user.has_perm("container.view_containerdistribution", distribution):
+            return True
+
+        return False
+
+    def has_push_permissions(self, path):
+        """
+        Check if the user has permissions to push to the repository specified by the path.
+        """
+        if not self.user.has_perm("container.add_containerdistribution"):
```

Comment:
https://github.com/pulp/pulp_container/pull/203#discussion_r557511475
This seems to rule out access a bit too early. (Basically meaning, that you need both Class-level create permissions.)

Reply:
The comment you have referenced is highlighting different commands. What do you mean by the class-level create permissions? Do we have such a thing? We consider either model-level permissions or object-level permissions. Since the object does not exist yet, we need to check for this model-level permission, right?

Reply:
Yeah, i meant model-level permission. But yes, the repository might already exist, when you attempt to push to it.
The model-level permission is to be considered, if it did not exist before, or you do not have explicit object permissions.

Reply:
this seems  correct to me - if a user does not have model level perm for adding distributions that means he cannot push at all, regardless into which namespace.

Reply:
but he could have object level permissions to push to a distribution/push_repo. And we completely rule this out here.

Reply:
we are removing post for push-repos

Reply:
are you confusing this with namespace/distribution? i might not following you

Reply:
given the name of this function, i believe it is questioning, whether a user can do a push operation on a specific path.
And whenever a distribution exists at that path, we should allow any user with object level "push_permission" (whatever that means in detail) for that distribution to push.
Only if there is not yet such a distribution, we need to ask if it's ok to create one.

Reply:
i see what you're saying. can we make any neede changes in the follow up PR? I would rather merge this as is to unblock folks

Reply:
sure.